sirepo/package_data/static/js/flash.js

```diff
@@ -131,6 +132,17 @@ SIREPO.app.controller('VisualizationController', function (appState, flashServic
                 frameCache.setFrameCount(data.frameCount, m);
             });
         }
+        if (! self.gridEvolutionColumnsSet && data.gridEvolutionColumns) {
+            self.gridEvolutionColumnsSet = true;
+            SIREPO.APP_SCHEMA.enum.GridEvolutionColumn = [];
+            for (var i = 0; i < data.gridEvolutionColumns.length; i++) {
+                var e = data.gridEvolutionColumns[i];
+                if (e[0] === '#') {
+                    continue;
+                }
+                SIREPO.APP_SCHEMA.enum.GridEvolutionColumn.push([e, e]);
```

Comment:
I think we had problems when we did this for job stuff. I think this probably needs to be on the server.

Reply:
I spoke with @moellep about this implementation. Maybe we misunderstood eachother. We do this sort of enum manipulation elsewhere (ex [in srw](https://github.com/radiasoft/sirepo/blob/9d7d346753a90486f6a77dd0e1c7c5c44b17c3a0/sirepo/package_data/static/js/srw.js#L85))

Can you elaborate a bit on what the on the server implementation would entail?

Reply:
Here's a commit where we had problems with jobRunMode changing: https://github.com/radiasoft/sirepo/commit/3484285758dc924b2781f721c78d38bca1a95f17

We eventually moved the dynamicity out of the schema for this reason.

Maybe this can't happen with FLASH or SRW.

Reply:
All of that code is still in the code base. What exactly did we move out (I can probably answer that question and will look more into it tomorrow)? My recollection of jobRunMode was that we had it hardcoded in the schema. That didn't work because the config could change and add/remove run modes. So we had to have the server dynamically send the available jobRunModes (on the authState). I think flash is doing the exact same thing (albeit at a different point in the lifecycle, status instead of authState, maybe that is the problem?).

Here is how I see it:
- jobRunMode: gets the list of [runModes from the server](https://github.com/radiasoft/sirepo/blob/09e28d2286e7cf0e7f608bd287eaedd8ebf0d589/sirepo/auth/__init__.py#L542) and populates them [on the enum](https://github.com/radiasoft/sirepo/blob/09e28d2286e7cf0e7f608bd287eaedd8ebf0d589/sirepo/package_data/static/js/sirepo.js#L144). 
- Flash: get the list of [column values from the server](https://github.com/radiasoft/sirepo/blob/09e28d2286e7cf0e7f608bd287eaedd8ebf0d589/sirepo/template/flash.py#L38) and populates them [on the enum](https://github.com/radiasoft/sirepo/blob/09e28d2286e7cf0e7f608bd287eaedd8ebf0d589/sirepo/package_data/static/js/flash.js#L143)

One problem I do see is in Flash is that jobRunMode has code for handling when the jobRunMode set by the user isn't known by the [gui or the server](https://github.com/radiasoft/sirepo/blob/09e28d2286e7cf0e7f608bd287eaedd8ebf0d589/sirepo/package_data/static/js/sirepo-components.js#L3325) or [just the server](https://github.com/radiasoft/sirepo/blob/09e28d2286e7cf0e7f608bd287eaedd8ebf0d589/sirepo/job_supervisor.py#L755). Flash doesn't have any of this code. So if the user has a column selected that is no longer available then the report generation code will fail and they will get back "report not generated." In that case should we choose a default value for them, send back a more helpful message, or keep report not generated?

Reply:
I guess I didn't remember how jobRunModeMap works. There's a difference in the way this works, for sure. 

JobRunMode does not change for a particular server. If it does, it's rarely done, and likely would be coupled with some other release so the GUI would refresh. 

GridEvolutionColumn is being changed as the result of a simulation. This is modifying global state in the app (the schema) based on user data that is generated dynamically for a particular simulation. I don't know what the values are, but could they possibly be added dynamically in a list, e.g. like lib files are?

Definitely need to handle "not there" cases. We want to know if it happens, at a minimum.

Reply:
https://github.com/radiasoft/sirepo/pull/2676/commits/6867d3d5b0cd825d636f681fbe341e7967026ceb